pkg/hooks/hooks.go

```diff
@@ -94,29 +96,83 @@ func runDirectory(dirpath string, environment []string, logger logging.Logger) e
 		if f.IsDir() {
 			continue
 		}
-		logger.WithField("path", fullpath).Infof("Executing hook %s", f.Name())
-		cmd := exec.Command(fullpath)
-		hookOut := &bytes.Buffer{}
-		cmd.Stdout = hookOut
-		cmd.Stderr = hookOut
-		cmd.Env = environment
-		err := cmd.Run()
-		if err != nil {
-			logger.WithErrorAndFields(err, logrus.Fields{
-				"path":   fullpath,
-				"output": hookOut.String(),
-			}).Warnf("Could not execute hook %s", f.Name())
-		} else {
-			logger.WithFields(logrus.Fields{
-				"path":   fullpath,
-				"output": hookOut.String(),
-			}).Debugln("Executed hook")
+
+		timeout := 5
+		h := Hook{fullpath, f.Name(), environment, logger, timeout}
+
+		err := h.RunWithTimeout()
+		htErr, ok := err.(HookTimeoutError)
+		if !ok {
+			logger.WithField("path", h.Path).Warnln(htErr.Error())
 		}
 	}
 
 	return nil
 }
 
+type Hook struct {
+	Path    string   // path to hook's executable
+	Name    string   // human-readable name of Hook
+	env     []string // unix environment in which the hook should be executed
+	logger  logging.Logger
+	timeout int // number of seconds before hook's execution times out
+}
+
+// HookTimeoutError is returned when a Hook's execution times out
+type HookTimeoutError struct {
+	H Hook
```

Comment:
can this simply be `Hook`?

When accessing the member you would have to use `Hook` instead of `H`, but perhaps good for clarity, and it is not such a big loss